File: packages/js/src/filter.ts
/**
 * Catalogue filtering: a plain options object and a chainable builder, which
 * both produce the same predicate.
 *
 * Within one field the values are ORed, across fields they are ANDed, exactly
 * as the ngano HTTP API behaves.
 *
 * @packageDocumentation
 */
import { resolveLanguages } from "./languages.js";
import type { Dataset } from "./types.js";

/**
 * Filter options accepted by {@link Catalogue.datasets} and {@link load}.
 * Every field is optional, and a field that is absent places no constraint.
 */
export interface FilterOptions {
  /** Free-text search over name, id, notes, languages, countries and host. */
  q?: string;
  /** ngano catalogue id or ids. */
  id?: string | string[];
  /**
   * Language tag, bare ISO 639-3 code or name. Every value resolves through the
   * ISO 639-3 registry to canonical tags, and a dataset matches when it carries
   * one of them. Unresolvable input matches nothing rather than everything.
   */
  language?: string | string[];
  /** The code-shaped spelling of {@link FilterOptions.language}. Same filter. */
  iso?: string | string[];
  /**
   * Widens a bare code to its regional varieties, so `eng` also matches
   * `eng-NG`. Off by default.
   */
  includeVarieties?: boolean;
  /** Country as ISO 3166-1 alpha-2 or as a name. */
  country?: string | string[];
  /** Region name, for example East Africa. */
  region?: string | string[];
  /** Task, for example ASR, TTS or ASR+TTS. */
  task?: string | string[];
  /** Language variety grouping. */
  variety?: string | string[];
  /**
   * `true` keeps only datasets whose commercial field is exactly "Yes". With
   * {@link FilterOptions.includePurchasable} it also keeps "Yes, if purchased".
   * `false` keeps only "No". A string or list matches the raw values.
   */
  commercial?: boolean | string | string[];
  /** Widens `commercial: true` to include "Yes, if purchased". */
  includePurchasable?: boolean;
  /** Licence family, for example "Attribution (CC-BY)". */
  licenceClass?: string | string[];
  /** Access route, for example Open, Request or Paid. */
  access?: string | string[];
  /** Transcription status, for example Transcribed. */
  labelled?: string | string[];
  /** Audio quality band. */
  quality?: string | string[];
  /** Recording domain. */
  domain?: string | string[];
  /** Host, for example HuggingFace. */
  host?: string | string[];
  /** Keeps only datasets with a Hugging Face repo, which are the loadable ones. */
  hfOnly?: boolean;
  /** Minimum published hours, inclusive. */
  minHours?: number;
  /** Maximum published hours, inclusive. */
  maxHours?: number;
  /** Keeps only datasets with a published hours figure. */
  hasHours?: boolean;
}

/**
 * Normalises a string for comparison: trimmed, lower case, punctuation folded.
 * Nullish input folds to an empty string, since catalogue fields are nullable.
 *
 * @param value any string, or nothing
 * @returns the comparison key
 */
export function fold(value: string | null | undefined): string {
  if (value === null || value === undefined) return "";
  return value
    .normalize("NFKD")
    .replace(/[̀-ͯ]/g, "")
    .toLowerCase()
    .replace(/[\s_\-'’.()/]+/g, "");
}

/**
 * Wraps a scalar or list option into an array of comparison keys.
 *
 * @param value one value, several values, or nothing
 * @returns folded values, empty when the option was absent
 */
function keys(value: string | string[] | undefined): string[] {
  if (value === undefined) return [];
  const list = Array.isArray(value) ? value : String(value).split(",");
  return list.map((entry) => fold(entry)).filter((entry) => entry.length > 0);
}

/**
 * Splits a scalar, list or comma-separated option into trimmed entries, left
 * exactly as typed. Language values are resolved rather than folded, so they
 * must not be stripped of their punctuation first.
 *
 * @param value one value, several values, or nothing
 * @returns the entries, empty when the option was absent
 */
function rawList(value: string | string[] | undefined): string[] {
  if (value === undefined) return [];
  const list = Array.isArray(value) ? value : String(value).split(",");
  return list.map((entry) => entry.trim()).filter((entry) => entry.length > 0);
}

/**
 * True when any of the dataset's values matches any of the wanted keys.
 *
 * @param wanted folded keys from the filter
 * @param values dataset values, unfolded
 * @returns whether the field passes
 */
function anyMatch(wanted: string[], values: Array<string | null | undefined>): boolean {
  if (wanted.length === 0) return true;
  const have = new Set(values.filter((v): v is string => typeof v === "string").map(fold));
  return wanted.some((key) => have.has(key));
}

/** Commercial values that count as usable in a commercial product. */
const COMMERCIAL_YES = "Yes";
/** Commercial value for datasets that allow commercial use once bought. */
const COMMERCIAL_PURCHASABLE = "Yes, if purchased";
/** Every value the catalogue uses for commercial use. */
const COMMERCIAL_VALUES = [COMMERCIAL_YES, COMMERCIAL_PURCHASABLE, "No", "Unstated"];

/**
 * Folds commercial values without splitting "Yes, if purchased" on its comma,
 * while still accepting comma-separated lists of the other values.
 *
 * @param values raw values from the caller
 * @returns folded keys
 */
function commercialKeys(values: string[]): string[] {
  const known = new Map(COMMERCIAL_VALUES.map((value) => [fold(value), value]));
  const out: string[] = [];
  for (const value of values) {
    const whole = fold(value);
    if (known.has(whole)) out.push(whole);
    else out.push(...keys(value));
  }
  return out;
}

/**
 * Builds a predicate from filter options.
 *
 * @param options the filter options
 * @returns a predicate over datasets
 */
export function compileFilter(options: FilterOptions = {}): (dataset: Dataset) => boolean {
  const ids = keys(options.id);
  // `language` and `iso` are one filter: both resolve through the registry.
  const requestedLanguages = [...rawList(options.language), ...rawList(options.iso)];
  const languageTags =
    requestedLanguages.length > 0
      ? new Set(
          resolveLanguages(requestedLanguages, {
            includeVarieties: options.includeVarieties === true,
          }).map((tag) => tag.toLowerCase()),
        )
      : null;
  const countries = keys(options.country);
  const regions = keys(options.region);
  const tasks = keys(options.task);
  const varieties = keys(options.variety);
  const licenceClasses = keys(options.licenceClass);
  const accesses = keys(options.access);
  const labelled = keys(options.labelled);
  const qualities = keys(options.quality);
  const domains = keys(options.domain);
  const hosts = keys(options.host);
  const query = options.q ? fold(options.q) : null;
  const rawQuery = options.q ? options.q.trim().toLowerCase() : null;

  let commercialValues: string[] | null = null;
  if (typeof options.commercial === "boolean") {
    commercialValues = options.commercial
      ? options.includePurchasable
        ? [COMMERCIAL_YES, COMMERCIAL_PURCHASABLE]
        : [COMMERCIAL_YES]
      : ["No"];
  } else if (options.commercial !== undefined) {
    // Not split on commas here: "Yes, if purchased" carries one of its own.
    commercialValues = Array.isArray(options.commercial)
      ? options.commercial
      : [String(options.commercial)];
  } else if (options.includePurchasable) {
    commercialValues = [COMMERCIAL_YES, COMMERCIAL_PURCHASABLE];
  }
  const commercial = commercialValues ? commercialKeys(commercialValues) : [];

  return (dataset: Dataset): boolean => {
    if (!anyMatch(ids, [dataset.id])) return false;
    if (languageTags !== null) {
      if (!dataset.languageTags.some((tag) => languageTags.has(tag.toLowerCase()))) return false;
    }
    if (!anyMatch(countries, [...dataset.countryCodes, ...dataset.countries])) return false;
    if (!anyMatch(regions, dataset.regions)) return false;
    if (!anyMatch(tasks, [dataset.task])) return false;
    if (!anyMatch(varieties, [dataset.variety])) return false;
    if (!anyMatch(commercial, [dataset.commercial])) return false;
    if (!anyMatch(licenceClasses, [dataset.licenceClass, dataset.licence])) return false;
    if (!anyMatch(accesses, [dataset.access])) return false;
    if (!anyMatch(labelled, [dataset.labelled])) return false;
    if (!anyMatch(qualities, [dataset.quality])) return false;
    if (!anyMatch(domains, [dataset.domain])) return false;
    if (!anyMatch(hosts, [dataset.host])) return false;

    if (options.hfOnly && !dataset.hfRepo) return false;
    if (options.hasHours && dataset.hoursNum === null) return false;
    if (options.minHours !== undefined) {
      if (dataset.hoursNum === null || dataset.hoursNum < options.minHours) return false;
    }
    if (options.maxHours !== undefined) {
      if (dataset.hoursNum === null || dataset.hoursNum > options.maxHours) return false;
    }

    if (query !== null && rawQuery !== null) {
      const haystack = [
        dataset.name,
        dataset.id,
        dataset.notes,
        dataset.domain,
        dataset.host,
        dataset.hfRepo,
        dataset.licence,
        ...dataset.languages,
        ...dataset.languagesClean,
        ...dataset.languageTags,
        ...dataset.countries,
        ...dataset.regions,
      ]
        .filter((value): value is string => typeof value === "string")
        .join(" ");
      if (!haystack.toLowerCase().includes(rawQuery) && !fold(haystack).includes(query)) {
        return false;
      }
    }

    return true;
  };
}

/**
 * Chainable filter builder. Every method returns the same instance, and
 * {@link Filter.toOptions} hands back a plain options object that can be passed
 * anywhere options are accepted.
 *
 * @example
 * ```ts
 * const f = new Filter().language("Shona").country("ZW").commercial(true).hfOnly();
 * catalogue.datasets(f);
 * ```
 */
export class Filter {
  private readonly options: FilterOptions = {};

  /**
   * Starts a new filter, optionally seeded from options.
   *
   * @param options initial options
   * @returns a new builder
   */
  static from(options: FilterOptions = {}): Filter {
    const filter = new Filter();
    Object.assign(filter.options, options);
    return filter;
  }

  /**
   * Free-text search.
   *
   * @param text the search text
   * @returns this builder
   */
  q(text: string): this {
    this.options.q = text;
    return this;
  }

  /**
   * Restricts to catalogue ids.
   *
   * @param values one or more ids
   * @returns this builder
   */
  id(...values: string[]): this {
    this.options.id = values;
    return this;
  }

  /**
   * Restricts to languages. Each value may be a BCP 47 tag, a bare ISO 639-3
   * code or any catalogue spelling of the name.
   *
   * @param values one or more tags, codes or names
   * @returns this builder
   */
  language(...values: string[]): this {
    this.options.language = values;
    return this;
  }

  /**
   * The code-shaped spelling of {@link Filter.language}, kept as an alias of it.
   *
   * @param values one or more ISO 639-3 codes or tags
   * @returns this builder
   */
  iso(...values: string[]): this {
    this.options.iso = values;
    return this;
  }

  /**
   * Widens a bare ISO 639-3 code to its regional varieties, so `eng` also
   * matches `eng-NG`.
   *
   * @param value whether to widen
   * @returns this builder
   */
  includeVarieties(value = true): this {
    this.options.includeVarieties = value;
    return this;
  }

  /**
   * Restricts to countries, by ISO 3166-1 alpha-2 code or name.
   *
   * @param values one or more countries
   * @returns this builder
   */
  country(...values: string[]): this {
    this.options.country = values;
    return this;
  }

  /**
   * Restricts to regions.
   *
   * @param values one or more region names
   * @returns this builder
   */
  region(...values: string[]): this {
    this.options.region = values;
    return this;
  }

  /**
   * Restricts to tasks.
   *
   * @param values for example "ASR" or "TTS"
   * @returns this builder
   */
  task(...values: string[]): this {
    this.options.task = values;
    return this;
  }

  /**
   * Restricts to language varieties.
   *
   * @param values for example "Indigenous"
   * @returns this builder
   */
  variety(...values: string[]): this {
    this.options.variety = values;
    return this;
  }

  /**
   * Restricts by commercial use. `true` means exactly "Yes".
   *
   * @param value true, false, or raw catalogue values
   * @returns this builder
   */
  commercial(value: boolean | string | string[] = true): this {
    this.options.commercial = value;
    return this;
  }

  /**
   * Widens `commercial(true)` to include "Yes, if purchased".
   *
   * @param value whether purchasable datasets are acceptable
   * @returns this builder
   */
  includePurchasable(value = true): this {
    this.options.includePurchasable = value;
    return this;
  }

  /**
   * Restricts to licence families.
   *
   * @param values for example "Attribution (CC-BY)"
   * @returns this builder
   */
  licenceClass(...values: string[]): this {
    this.options.licenceClass = values;
    return this;
  }

  /**
   * Restricts to access routes.
   *
   * @param values for example "Open"
   * @returns this builder
   */
  access(...values: string[]): this {
    this.options.access = values;
    return this;
  }

  /**
   * Restricts by transcription status.
   *
   * @param values for example "Transcribed"
   * @returns this builder
   */
  labelled(...values: string[]): this {
    this.options.labelled = values;
    return this;
  }

  /**
   * Restricts by audio quality band.
   *
   * @param values for example "Standard (16 kHz)"
   * @returns this builder
   */
  quality(...values: string[]): this {
    this.options.quality = values;
    return this;
  }

  /**
   * Restricts by recording domain.
   *
   * @param values for example "Broadcast news"
   * @returns this builder
   */
  domain(...values: string[]): this {
    this.options.domain = values;
    return this;
  }

  /**
   * Restricts by host.
   *
   * @param values for example "HuggingFace"
   * @returns this builder
   */
  host(...values: string[]): this {
    this.options.host = values;
    return this;
  }

  /**
   * Keeps only datasets that have a Hugging Face repo.
   *
   * @param value whether to apply the restriction
   * @returns this builder
   */
  hfOnly(value = true): this {
    this.options.hfOnly = value;
    return this;
  }

  /**
   * Sets a lower bound on published hours.
   *
   * @param hours minimum hours, inclusive
   * @returns this builder
   */
  minHours(hours: number): this {
    this.options.minHours = hours;
    return this;
  }

  /**
   * Sets an upper bound on published hours.
   *
   * @param hours maximum hours, inclusive
   * @returns this builder
   */
  maxHours(hours: number): this {
    this.options.maxHours = hours;
    return this;
  }

  /**
   * Keeps only datasets with a published hours figure.
   *
   * @param value whether to apply the restriction
   * @returns this builder
   */
  hasHours(value = true): this {
    this.options.hasHours = value;
    return this;
  }

  /**
   * Hands back the accumulated options.
   *
   * @returns a shallow copy of the options
   */
  toOptions(): FilterOptions {
    return { ...this.options };
  }

  /**
   * Tests one dataset against the filter.
   *
   * @param dataset the dataset to test
   * @returns whether it passes
   */
  match(dataset: Dataset): boolean {
    return compileFilter(this.options)(dataset);
  }
}

/**
 * Accepts either a {@link Filter} or a plain options object.
 */
export type FilterLike = Filter | FilterOptions;

/**
 * Reduces a filter-like value to plain options.
 *
 * @param filter a builder, options, or nothing
 * @returns plain options
 */
export function toOptions(filter?: FilterLike): FilterOptions {
  if (!filter) return {};
  return filter instanceof Filter ? filter.toOptions() : filter;
}

File: packages/rust/src/filter.rs
//! Dataset selection.

use crate::language::LanguageRegistry;
use crate::model::Dataset;

/// A declarative dataset filter.
///
/// Values given to the same method are combined with OR, and different methods
/// are combined with AND. Every string comparison is case-insensitive.
///
/// ```
/// use ngano::{Catalogue, Filter};
///
/// let cat = Catalogue::bundled()?;
/// let f = Filter::new().language("sna").commercial(true).task("ASR");
/// for d in cat.datasets(&f) {
///     assert!(d.allows_commercial_use());
/// }
/// # Ok::<(), ngano::NganoError>(())
/// ```
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Filter {
    /// Free-text query applied to name, id, languages, countries and notes.
    pub q: Option<String>,
    /// Languages to keep, each written as a bare ISO 639-3 code, a BCP 47 tag
    /// or a name. Every value is resolved to canonical tags and matched against
    /// [`Dataset::language_tags`].
    pub languages: Vec<String>,
    /// Widen a bare ISO 639-3 code to every regional variety of it, so `eng`
    /// also matches `eng-NG` and `eng-ZA`.
    pub include_varieties: bool,
    /// Countries, given as ISO 3166-1 alpha-2 codes or as names.
    pub countries: Vec<String>,
    /// Regional groupings.
    pub regions: Vec<String>,
    /// Tasks, for example `ASR`.
    pub tasks: Vec<String>,
    /// Language varieties.
    pub varieties: Vec<String>,
    /// Licence families.
    pub licence_classes: Vec<String>,
    /// Access modes.
    pub access: Vec<String>,
    /// Labelling states.
    pub labelled: Vec<String>,
    /// Quality bands.
    pub quality: Vec<String>,
    /// Recording domains.
    pub domains: Vec<String>,
    /// Hosts, for example `HuggingFace`.
    pub hosts: Vec<String>,
    /// `Some(true)` keeps datasets usable commercially (`Yes` or
    /// `Yes, if purchased`), `Some(false)` keeps only those that are not.
    pub commercial: Option<bool>,
    /// `Some(true)` keeps only datasets with a Hugging Face repo.
    pub hf_only: Option<bool>,
    /// `Some(true)` keeps only datasets with a numeric hours figure.
    pub has_hours: Option<bool>,
    /// Lower bound on countable hours, inclusive.
    pub min_hours: Option<f64>,
    /// Upper bound on countable hours, inclusive.
    pub max_hours: Option<f64>,
    /// When true, drop records whose size is self-reported and unverified.
    pub exclude_unverified: bool,
}

macro_rules! push_setter {
    ($(#[$m:meta])* $name:ident, $field:ident) => {
        $(#[$m])*
        #[must_use]
        pub fn $name(mut self, value: impl Into<String>) -> Self {
            self.$field.push(value.into());
            self
        }
    };
}

impl Filter {
    /// An empty filter, which matches every record.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Free-text query over name, id, languages, countries, domain and notes.
    #[must_use]
    pub fn q(mut self, value: impl Into<String>) -> Self {
        self.q = Some(value.into());
        self
    }

    push_setter!(
        /// Add a language to match, written as a bare ISO 639-3 code (`sna`),
        /// a BCP 47 tag (`eng-NG`) or a name (`Shona`, `isiZulu`). Case does
        /// not matter, and an unresolvable value simply matches nothing.
        language, languages);
    push_setter!(
        /// Add a language by its ISO 639-3 code. This is the code-shaped
        /// spelling of [`Filter::language`] and resolves identically.
        iso, languages);
    push_setter!(
        /// Add a country, as an ISO 3166-1 alpha-2 code or a name.
        country, countries);
    push_setter!(
        /// Add a regional grouping to match.
        region, regions);
    push_setter!(
        /// Add a task to match, for example `ASR`.
        task, tasks);
    push_setter!(
        /// Add a language variety to match.
        variety, varieties);
    push_setter!(
        /// Add a licence family to match.
        licence_class, licence_classes);
    push_setter!(
        /// Add an access mode to match, for example `Open`.
        access_mode, access);
    push_setter!(
        /// Add a labelling state to match, for example `Transcribed`.
        labelled, labelled);
    push_setter!(
        /// Add a quality band to match.
        quality, quality);
    push_setter!(
        /// Add a recording domain to match.
        domain, domains);
    push_setter!(
        /// Add a host to match, for example `HuggingFace`.
        host, hosts);

    /// Keep datasets by commercial usability. `true` keeps `Yes` and
    /// `Yes, if purchased`, `false` keeps everything else.
    #[must_use]
    pub fn commercial(mut self, value: bool) -> Self {
        self.commercial = Some(value);
        self
    }

    /// Widen every bare ISO 639-3 code given to [`Filter::language`] to its
    /// regional varieties, so `eng` also keeps `eng-NG` and `eng-ZA`. Off by
    /// default: a bare code means the language at large and must never silently
    /// pick up a country-specific variety.
    ///
    /// ```
    /// use ngano::{Catalogue, Filter};
    ///
    /// let cat = Catalogue::bundled()?;
    /// let plain = cat.datasets(&Filter::new().language("eng"));
    /// let wide = cat.datasets(&Filter::new().language("eng").include_varieties(true));
    /// assert!(wide.len() > plain.len());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    #[must_use]
    pub fn include_varieties(mut self, value: bool) -> Self {
        self.include_varieties = value;
        self
    }

    /// Keep only datasets hosted on Hugging Face, which are the streamable ones.
    #[must_use]
    pub fn hf_only(mut self, value: bool) -> Self {
        self.hf_only = Some(value);
        self
    }

    /// Keep only datasets that publish a numeric hours figure.
    #[must_use]
    pub fn has_hours(mut self, value: bool) -> Self {
        self.has_hours = Some(value);
        self
    }

    /// Lower bound on countable hours, inclusive.
    #[must_use]
    pub fn min_hours(mut self, value: f64) -> Self {
        self.min_hours = Some(value);
        self
    }

    /// Upper bound on countable hours, inclusive.
    #[must_use]
    pub fn max_hours(mut self, value: f64) -> Self {
        self.max_hours = Some(value);
        self
    }

    /// Drop records whose published size is self-reported and unverified.
    #[must_use]
    pub fn exclude_unverified(mut self, value: bool) -> Self {
        self.exclude_unverified = value;
        self
    }

    /// The canonical BCP 47 tags this filter keeps, deduplicated and in the
    /// order the languages were added. A value that resolves to nothing
    /// contributes nothing, so a filter written only of unknown languages keeps
    /// no records at all.
    ///
    /// ```
    /// use ngano::Filter;
    ///
    /// let f = Filter::new().language("isiZulu").iso("sna");
    /// assert_eq!(f.resolved_languages(), vec!["zul".to_string(), "sna".to_string()]);
    /// ```
    pub fn resolved_languages(&self) -> Vec<String> {
        LanguageRegistry::shared().resolve_all(&self.languages, self.include_varieties)
    }

    /// Languages given to this filter that the registry does not recognise.
    /// Useful for telling a caller that a spelling matched nothing, rather than
    /// silently returning an empty result.
    pub fn unresolved_languages(&self) -> Vec<String> {
        let registry = LanguageRegistry::shared();
        self.languages
            .iter()
            .filter(|v| registry.resolve(v, self.include_varieties).is_empty())
            .cloned()
            .collect()
    }

    /// True when this filter would keep every record.
    pub fn is_empty(&self) -> bool {
        *self == Filter::default()
    }

    /// Test one record against the filter.
    ///
    /// ```
    /// use ngano::{Catalogue, Filter};
    ///
    /// let cat = Catalogue::bundled()?;
    /// let d = cat.all().first().expect("catalogue is not empty");
    /// assert!(Filter::new().matches(d));
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn matches(&self, d: &Dataset) -> bool {
        if self.exclude_unverified && d.unverified_size {
            return false;
        }
        if let Some(want) = self.commercial {
            if d.allows_commercial_use() != want {
                return false;
            }
        }
        if self.hf_only == Some(true) && !d.is_streamable() {
            return false;
        }
        if self.hf_only == Some(false) && d.is_streamable() {
            return false;
        }
        if let Some(want) = self.has_hours {
            if d.hours_num.is_some() != want {
                return false;
            }
        }
        if self.min_hours.is_some() || self.max_hours.is_some() {
            let Some(h) = d.countable_hours() else {
                return false;
            };
            if self.min_hours.is_some_and(|m| h < m) || self.max_hours.is_some_and(|m| h > m) {
                return false;
            }
        }
        if !any_match(&self.tasks, std::slice::from_ref(&d.task))
            || !any_match(&self.varieties, std::slice::from_ref(&d.variety))
            || !any_match(
                &self.licence_classes,
                std::slice::from_ref(&d.licence_class),
            )
            || !any_match(&self.access, std::slice::from_ref(&d.access))
            || !any_match(&self.labelled, std::slice::from_ref(&d.labelled))
            || !any_match(&self.quality, std::slice::from_ref(&d.quality))
            || !any_match(&self.domains, std::slice::from_ref(&d.domain))
            || !any_match(&self.hosts, std::slice::from_ref(&d.host))
            || !any_match(&self.regions, &d.regions)
        {
            return false;
        }
        if !self.languages.is_empty() {
            let wanted = self.resolved_languages();
            if !d
                .language_tags
                .iter()
                .any(|tag| wanted.iter().any(|w| w.eq_ignore_ascii_case(tag)))
            {
                return false;
            }
        }
        if !any_match_either(&self.countries, &d.country_codes, &d.countries) {
            return false;
        }
        if let Some(q) = &self.q {
            if !matches_query(d, q) {
                return false;
            }
        }
        true
    }
}

/// True when `wanted` is empty, or when any wanted value equals any candidate,
/// ignoring case.
fn any_match(wanted: &[String], candidates: &[String]) -> bool {
    if wanted.is_empty() {
        return true;
    }
    wanted.iter().any(|w| {
        candidates
            .iter()
            .any(|c| c.eq_ignore_ascii_case(w.trim()) || c.trim().eq_ignore_ascii_case(w.trim()))
    })
}

/// True when `wanted` is empty, or when it matches either candidate list.
fn any_match_either(wanted: &[String], first: &[String], second: &[String]) -> bool {
    if wanted.is_empty() {
        return true;
    }
    any_match(wanted, first) || any_match(wanted, second)
}

/// Case-insensitive substring search across the fields a human would search.
fn matches_query(d: &Dataset, q: &str) -> bool {
    let needle = q.trim().to_lowercase();
    if needle.is_empty() {
        return true;
    }
    let mut haystack = String::with_capacity(256);
    haystack.push_str(&d.name);
    haystack.push('\n');
    haystack.push_str(&d.id);
    haystack.push('\n');
    haystack.push_str(&d.domain);
    haystack.push('\n');
    haystack.push_str(&d.notes);
    haystack.push('\n');
    if let Some(r) = &d.hf_repo {
        haystack.push_str(r);
        haystack.push('\n');
    }
    for list in [
        &d.languages,
        &d.languages_clean,
        &d.language_tags,
        &d.countries,
        &d.regions,
    ] {
        for v in list {
            haystack.push_str(v);
            haystack.push('\n');
        }
    }
    haystack.to_lowercase().contains(&needle)
}
